Test package installation in temporary virtual environment

matt-graham opened the pull request:
Resolves #516 

Creates a virtual environment in temporary package directory and tests package installation in there rather than installing in parent environment tests are run from.

Calling `pip install` command by directly referencing Python executable in created virtual environment seems to work in terms of ensuring package is installed in `site-packages` of temporary virtual environment rather than parent environment (at least I no longer see a `cookiecutter-test` entry when running `pip list` after running tests) and it seemed non trivial to `source` the virtual environment script with `subprocess.run`.

matt-graham committed:
Test package installation in temporary virtual environment

matt-graham committed:
Use Windows specific venv path

matt-graham commented:
I don't think the link checker check fail is anything to do with changes in this PR as it doesn't change or add any URLs. It looks like it might be due to
```
[304] https://discord.com/ | Redirect (304 Not Modified): Not Modified
```
flagging `https://discord.com/` as a redirect, potentially because of the trailing slash?

samcunliffe reviewed (commented):
Looks great!

2 ± 1 things come to mind:
* Did you look at [`pytest-venv`](https://github.com/mmerickel/pytest-venv)? It's unfortunately not receiving so much dev (last commit ~2 years ago). But it (or something similar) might cut down on some of the boilerplate?
* Alternatively we could plop all of this into a fixture (i.e. do as they do in the `pytest-venv` plugin but with our own code.
* The whole package generation test... um is there a smarter way to do it (I appreciate that is less of a comment for this bugfix and more of a general unhelpful musing).

matt-graham commented:
>  * Did you look at [`pytest-venv`](https://github.com/mmerickel/pytest-venv)? It's unfortunately not receiving so much dev (last commit ~2 years ago). But it (or something similar) might cut down on some of the boilerplate?
>  * Alternatively we could plop all of this into a fixture (i.e. do as they do in the `pytest-venv` plugin but with our own code.

I hadn't come across `pytest-venv` and my cursory searching for `pytest temporary virtual environment` didn't throw it up, but it does look like it does exactly what we want including dealing with the Windows specific directory layout that caught me out. While the lack of recent commits is a bit of a flag, it is a very minimal package so can imagine there isn't too much that can break and so need maintenance. I'd be happy to add it as a test dependency rather than reinventing the wheel ourselves. Any opinions to the contrary?

paddyroddy commented:
> The whole package generation test... um is there a smarter way to do it

There must be. I do think it's currently causing more problems than it has solved.

Saransh-cpp commented:
There is https://github.com/hackebrot/pytest-cookies, but it is unmaintained as well (tho used for testing in https://github.com/osl-incubator/scicookie).

If we end up migrating to copier, we can use https://github.com/12rambau/pytest-copie.

matt-graham committed:
Using pytest-venv rather than manually creating temp venv

matt-graham commented:
Updated to use `pytest_venv` fixture rather than manual solution. `mypy` was complaining that
```
error: Cannot find implementation or library stub for module named "pytest_venv"  [import-not-found]
```
so I added a manual override to ignore import. Alternative would be not to add a type annotation for `venv` argument and add a `noqa` comment to stop Ruff complaining about this.

matt-graham commented:
Link checker flaking again (tried one re-run) this time due to
```
[503] https://coveralls.io/pricing | Network error: Service Unavailable
```

paddyroddy commented:
> Link checker flaking again (tried one re-run) this time due to
> 
> ```
> [503] https://coveralls.io/pricing | Network error: Service Unavailable
> ```

Coveralls is currently down

samcunliffe reviewed (approved):
Nice!

I'd say the new dependency is worth it for the one-liner environmentisation of our messy tests. (I think I've invented "environmentisation".)

paddyroddy commented on tests/test_package_generation.py:
```diff
@@ -8,6 +8,7 @@
 import typing
 
 import pytest
+import pytest_venv
```
What about this one-liner instead?
```suggestion
import pytest_venv  # type: ignore[import-not-found]
```

paddyroddy commented on pyproject.toml:
```diff
@@ -1,3 +1,7 @@
+[[tool.mypy.overrides]]
+ignore_missing_imports = true
+module = ["pytest_venv"]
```
If we keep this, then I'd vote for splitting like this like the rest of the file. Means shorter diffs if we add to it and consistency.
```suggestion
module = [
	"pytest_venv",
]
```

matt-graham replied:
Have went with suggestion for inline ignore as seemed cleaner so marking this as resolved.

paddyroddy reviewed (commented):
Nice, `pytest_venv` looks much neater than previous iteration

Saransh-cpp reviewed (approved):
Nice! 🚀

matt-graham committed:
Update tests/test_package_generation.py

Co-authored-by: Patrick J. Roddy <patrickjamesroddy@gmail.com>

matt-graham committed:
Removing now redundant mypy config

matt-graham committed:
Merge branch 'main' into mmg/use-venv-in-install-test
